### PR TITLE
Pager with address sanitizer

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: core/arch/arm/kernel/generic_boot.c [hunk 1]
```diff
@@ -280,14 +311,18 @@ static void init_runtime(unsigned long pageable_part)
...
 
 	/*
 	 * Need tee_mm_sec_ddr initialized to be able to allocate secure
 	 * DDR below.
 	 */
 	teecore_init_ta_ram();
 
+#ifdef CFG_CORE_SANITIZE_KADDRESS
+	carve_out_asan_mem(&tee_mm_sec_ddr);
+#endif
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_boot.c]: minor: suggest to remove the `#ifdef` and use a empty definition.

File: core/arch/arm/kernel/generic_boot.c [hunk 2]
```diff
@@ -304,6 +304,7 @@ static void init_runtime(unsigned long pageable_part)
 
 	thread_init_boot_thread();
 
+	init_asan();
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_boot.c]: minor: suggest an empty line below.

File: core/arch/arm/kernel/generic_boot.c [hunk 3]
```diff
@@ -223,6 +224,17 @@ static void init_asan(void)
 	asan_tag_access(&__initcall_start, &__initcall_end);
 	asan_tag_access(&__ctor_list, &__ctor_end);
 	asan_tag_access(__rodata_start, __rodata_end);
+#ifdef CFG_WITH_PAGER
+#ifdef CFG_CORE_RODATA_NOEXEC
+	asan_tag_access((void *)VCORE_INIT_RO_PA,
+			(void *)(VCORE_INIT_RO_PA + VCORE_INIT_RO_SZ));
+	asan_tag_access(__rodata_pageable_start, __rodata_pageable_end);
+#else
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/generic_boot.c]: Hmm, it seems everything could be reduced to
`asan_tag_access(__pageable_start, __pageable_end);`

**Comment by jenswi-linaro** [comment]
Comments addressed, some discovered problems addressed.

File: core/arch/arm/kernel/generic_boot.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/kernel/generic_boot.c]: ok when no RONOEXE. Else there are holes in due to rx/ro split.

File: core/arch/arm/kernel/generic_boot.c [hunk 4]
```diff
@@ -179,7 +180,120 @@ static void init_vfp_sec(void)
...
+	 * Add access to areas that aren't opened automatically by a
+	 * constructor.
+	 */
+	asan_tag_access(&__initcall_start, &__initcall_end);
+	asan_tag_access(&__ctor_list, &__ctor_end);
+	asan_tag_access(__rodata_start, __rodata_end);
+#ifdef CFG_WITH_PAGER
+	asan_tag_access(__pageable_start, __pageable_end);
+#endif /*CFG_WITH_PAGER*/
+	asan_tag_access(__nozi_start, __nozi_end);
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_boot.c]: note: this change discards the `asan_tag_access()` on exidx and extab sections. Seems unwinding related. These do not need a manual access tagging ?

File: core/arch/arm/kernel/generic_boot.c [hunk 5]
```diff
@@ -179,7 +180,120 @@ static void init_vfp_sec(void)
...
+			 * the pool.
+			 */
+			asz -= pool->lo - apa;
+			apa = pool->lo;
+		} else {
+			/*
+			 * ASAN buffer is overlapping with the end of the
+			 * pool.
+			 */
+			asz = apa - pool->hi;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_boot.c]: `asz = pool->hi - apa;`

File: core/arch/arm/kernel/generic_boot.c [hunk 6]
```diff
@@ -179,7 +180,120 @@ static void init_vfp_sec(void)
...
+		}
+	}
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_boot.c]: do we need to assert asan map is not fully overlapping the target pool ?   
`assert(apa >= pool.lo || (apa + asz) <= pool->hi);`
(unlikely to happen)

File: core/arch/arm/kernel/generic_boot.c [hunk 7]
```diff
@@ -179,7 +180,120 @@ static void init_vfp_sec(void)
...
+#else
+static void carve_out_asan_mem(tee_mm_pool_t *pool __unused)
+{
+}
+#endif
+
+static void init_vcore(tee_mm_pool_t *mm_vcore)
+{
+	vaddr_t begin = TEE_RAM_VA_START;
+	vaddr_t end =  TEE_RAM_VA_START + CFG_TEE_RAM_VA_SIZE;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_boot.c]: minor: trailing space. add `const` to `begin` definition.

File: core/arch/arm/kernel/generic_boot.c (hunk 4 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/generic_boot.c]: Thanks, I'll fix

**Comment by jenswi-linaro** [comment]
Comments addressed
Added a commit fixing `__asan_handle_no_return()`

File: core/arch/arm/mm/core_mmu.c [hunk 8]
```diff
@@ -97,14 +97,15 @@ register_phys_mem(MEM_AREA_TEE_RAM_RW, VCORE_UNPG_RW_PA, VCORE_UNPG_RW_SZ);
...
-#ifdef CFG_CORE_SANITIZE_KADDRESS
-register_phys_mem(MEM_AREA_TEE_ASAN, ASAN_MAP_PA, ASAN_MAP_SZ);
-#endif
 #endif /*CFG_WITH_PAGER*/
 #else /*!CFG_CORE_RWDATA_NOEXEC*/
 register_phys_mem(MEM_AREA_TEE_RAM, CFG_TEE_RAM_START, CFG_TEE_RAM_PH_SIZE);
 #endif /*!CFG_CORE_RWDATA_NOEXEC*/
 
+#if defined(CFG_CORE_SANITIZE_KADDRESS) && defined(CFG_WITH_PAGER)
+register_phys_mem(MEM_AREA_TEE_ASAN, ASAN_MAP_PA, ASAN_MAP_SZ);
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c]: A comment to state /* Asan ram is part of `MEM_AREA_TEE_RAM_RW` when pager is disable */

File: core/kernel/asan.c [hunk 9]
```diff
@@ -275,9 +275,8 @@ void __noreturn __asan_report_store_n_noabort(vaddr_t addr, size_t size)
 }
 
 void __asan_handle_no_return(void);
-void __noreturn __asan_handle_no_return(void)
+void __asan_handle_no_return(void)
 {
-	panic();
```
Comment (etienne-lms) [inline core/kernel/asan.c]: no kind of panicking ?

File: core/arch/arm/mm/core_mmu.c [hunk 10]
```diff
@@ -97,9 +97,13 @@ register_phys_mem(MEM_AREA_TEE_RAM_RW, VCORE_UNPG_RW_PA, VCORE_UNPG_RW_SZ);
...
 register_phys_mem(MEM_AREA_TEE_RAM_RO, VCORE_INIT_RO_PA, VCORE_INIT_RO_SZ);
-#endif
-#else
+#endif /*CFG_WITH_PAGER*/
+#else /*!CFG_CORE_RWDATA_NOEXEC*/
 register_phys_mem(MEM_AREA_TEE_RAM, CFG_TEE_RAM_START, CFG_TEE_RAM_PH_SIZE);
+#endif /*!CFG_CORE_RWDATA_NOEXEC*/
+
+#if defined(CFG_CORE_SANITIZE_KADDRESS) && defined(CFG_WITH_PAGER)
+register_phys_mem(MEM_AREA_TEE_ASAN, ASAN_MAP_PA, ASAN_MAP_SZ);
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c]: A comment to state `/* Asan ram is part of MEM_AREA_TEE_RAM_RW when pager is disable */` ?

File: core/kernel/asan.c (hunk 9 above)
Comment (jenswi-linaro) [reply core/kernel/asan.c]: No, it's not entirely clear to me what we're supposed to do here. The Linux kernel does nothing though.

**Comment by jenswi-linaro** [comment]
Update

**Comment by jenswi-linaro** [comment]
Ping?

**Comment by etienne-lms** [comment]
Typo in generic_boot.c.
Aside this and rebase requirement, `Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for "core: arm: kern.ld.S: put constructors in init", `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for the others.

**Comment by jenswi-linaro** [comment]
Rebased, tags applied

**Comment by jforissier** [comment]
There is one long line in the commit description of "qemu_virt: fix memory configuration". Can you fix it please?

**Comment by jenswi-linaro** [comment]
Updated

**Comment by jforissier** [comment]
Thanks

**Pull request merged**